### ISSUE TITLE
Support for AWS Lambda

### ISSUE THREAD

**Issue by joe153** [Netflix/metaflow#358]
What would it take for Metaflow to support @lambda decorator? AWS Batch is powerful and good at bigger jobs but AWS Lambda would be great for smaller/fast jobs and would work together with AWS Step Functions.

**Comment by savingoyal** [comment]
@joe153 For AWS Lambda support, you would have to implement [a plugin very similar to the `@batch` plugin](https://github.com/Netflix/metaflow/tree/master/metaflow/plugins/aws/batch). Job execution guarantees are going to be quite different on AWS Lambda because of the resource constraints and timing constraints on the Lambda containers. For example, `@conda` might not work because of the 3G limit on memory and 0.5G limit on disk; `Parameters` (and `foreaches`) could also be a bit problematic due to the 4K env size limit and it will be hard to gauge if the function invocation will always be within the 15 mins limit.

**Comment by Zylatis** [comment]
I think it's worth looking into. I was considering doing something like this independent of Metaflow as a kind of super lazy 'parallel this loop' up-and-out onto Lambda but it would be great to have in Metaflow. 

All the cautionary points above are valid, but I think if someone is explicitly invoking @lambda not @batch it's reasonable to assume they are aware of the limitations, and if they get timeouts it's on them. If there are embarrassingly parallel tasks which the user is certain will be < 15 min (and all the other issues discussed, env size etc) then this would be a powerful tool.

**Comment by calmacx** [comment]
hi all, I also agree it'd be fantastic to have this. @joe153 are you considering working implementing this plugin yourself? I might give it a go myself

**Comment by joe153** [comment]
@Zylatis: I agree with you. Metaflow syntax is super simple and neat, it would be great to generate step functions using Metaflow with Lambda. 3G memory limit and 15 timeout limit are enough for many jobs and they will only increase in the future. Yes Metaflow is great at ML projects but I see broader use cases like copy data, send emails, etc.

@calmacx I haven't had a chance to do anything. Give it a go!

**Comment by Zylatis** [comment]
@calmacx I'd be keen to collaborate if you're looking to take this on and it won't be too many cooks etc.

**Comment by calmacx** [comment]
Great stuff. When I get round to starting a fork I'll tag you guys... also, don't let me stop you making a start yourself, might be ~1 week or so before I get started. Very happy to collab - I'm a research scientist - my excuse if my coding isn't great...

**Comment by savingoyal** [comment]
Great! Keep me updated and let me know if you need any assistance grokking the codebase.

**Comment by calmacx** [comment]
hi guys,

FYI, I finally made a start on this late last week, and spent some time yesterday and today. I'm going to hopefully hash out something very hacky to get it working, and then we can take it from there (I'll tag anyone interested to see if it's worth salvaging) 

As I said, I'm a research scientist so hacking code to get it to work is what i do...  😄 

cheers,
Calum

**Comment by savingoyal** [comment]
The newly announced [container support for AWS Lambda](https://aws.amazon.com/blogs/aws/new-for-aws-lambda-container-image-support/) as well as [increased resource limits ](https://aws.amazon.com/blogs/aws/new-for-aws-lambda-functions-with-up-to-10-gb-of-memory-and-6-vcpus/) is quite timely, I must say :)

**Comment by calmacx** [comment]
ohh yes indeed! Im glad i spoke up 😅 

@savingoyal do you think the simplest way to go would be to just containerise everything? 10GB is pretty decent for the image size, and as pointed out by others, anyone using Lambda over Batch will be aware of the limits

Therefore, instead of trying to convert any step function into a (aws) lambda function, we could just containerise the datastore and code, following something similar to what that batch does, and run this on aws lambda..
```
 python -m pip install awscli click requests boto3 -qqq &&mkdir metaflow \
  && .... &&\
 python -u FOO.py --quiet .... --pylint step BAR --run-id <> --task-id <>  --input-paths ${METAFLOW_INPUT_PATHS_0} ..
```
I'm sure it could be made more efficient if you only containerised the necessary inputs too.

Is that the way you would recommend to approach it, or have other ideas? Seeing as I'm still struggling to see the global picture of how this would connect up to using AWS Step Functions too.

**Comment by savingoyal** [comment]
@calmacx For natively executing Metaflow tasks on AWS Lambda, we would have to build the functionality that closely mimics the [current AWS Batch implementation](https://github.com/Netflix/metaflow/tree/master/metaflow/plugins/aws/batch). Rather than containerize the code (if I understood the intent correctly), we can just launch the job directly on top of the [runtime base images for AWS Lambda](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-images.html). We can copy over the code and inputs as part of the actual function execution. 

The AWS Step Functions integration might be a bit of work because of some of the idiosyncrasies of Amazon State Language. One thought is to use AWS Lambda for quick development and use AWS Batch for production workloads.

**Comment by calmacx** [comment]
@savingoyal - thanks, I worded my message badly, that's what I've been doing (closely mimic AWS Batch).

So I was just meaning, we make new images on top of the runtime base images, [as is done in the original example](https://aws.amazon.com/blogs/aws/new-for-aws-lambda-container-image-support/), and is what is suggested to do. 

Anyway, better for me to gave a go and come back to you later 👍

**Comment by oavdeev** [comment]
So I've started working on an implementation of this, keeping track of progress in this draft PR #463 

I'm still not 100% sure if having something like `@fargate` instead would make more sense, there are some limitations in Lambda (mentioned here and in that PR) that could be annoying to work around.
